### DISCUSSION TITLE
useQuery vs new QueryClient inside Context

### DISCUSSION THREAD

**Discussion by Zerebokep** [TanStack/query#3937] [Q&A]
I was wondering if I can use useQuery inside a react context or are there any side-effects? I remember that I've read somewhere that it could lead to bugs, but not sure what the reason was.

Also it seems automatic refetching (refetchInterval) does not work inside a context when using new QueryClient().

**Comment by TkDodo** [comment]
What does "inside context" mean? Can you show an example?

**Reply by Zerebokep** [reply]
Sure:

```
export function AuthProvider({ children }: Props) {
  const [user, setUser] = useState()

  // useQuery() or new QueryClient()?

  const value = useMemo(
    () => ({
      user,
      setUser,
    }),
    [user, setUser]
  )

  return <Context.Provider value={value}>{children}</Context.Provider>
}
```

```
<QueryClientProvider client={queryClient}>
  <Hydrate state={pageProps.dehydratedState}>
      <AuthProvider>
        <Component {...pageProps} />
      </AuthProvider>
  </Hydrate>
  <ReactQueryDevtools initialIsOpen={false} />
</QueryClientProvider>
```

**Reply by TkDodo** [reply]
> // useQuery() or new QueryClient()?

`useQuery` and `new QueryClient()` are two _very_ different things. My guess is you want `useQuery` here, why would you do `new QueryClient()` ? What do you expect this to achieve?

**Reply by Zerebokep** [reply]
I was just wondering if useQuery is safe to use inside a context, because it uses (from my understanding) an existing queryclient, while new QueryClient().fetchQuery() doesn't.

**Reply by TkDodo** [reply]
1) Your `AuthProvider` is a normal, standard functional component. It doesn't really matter that it renders a context or not, that doesn't make react-query be "inside a context". I find that viewpoint of "inside a context" quite confusing tbh.

2) Inside a functional component, you can call `useQuery`, and yes, it expects to have the `QueryClientProvider` in the tree above it.

3) you cannot call `new QueryClient().fetchQuery()` during render, because it fires an api request and returns a promise. `new QueryClient()` also creates a new cache, so there's not caching if you do that. There are not examples or docs about calling `new QueryClient().fetchQuery` apart from SSR.

---

> // useQuery() or new QueryClient()?

so the answer is: `useQuery`.

**Reply by Zerebokep** [reply]
I see, thanks. I'm quite new to react, so please excuse the dumb questions. I currently also use fetchQuery inside callbacks like useEffect or zod js validation schemes (because of [this](https://reactjs.org/docs/hooks-rules.html)), is there also a way to use useQuery here?

**Reply by kxxoling** [reply]
useQuery is another react hook, of course you can use it in context as if it's wrapped in QueryClientProvider.
